core/communications/httpCommunication.go

```diff
@@ -1054,145 +1417,349 @@ func (communication *HTTP) handlePutData(orgID string, objectType string, object
 	} else if notification.InstanceID != instanceID {
 		if log.IsLogging(logger.ERROR) {
 			log.Error("Handling object get data, notification.InstanceID(%d) != metaData,InstanceID(%d), notification status(%s) for %s %s %s %s %s\n", notification.InstanceID, instanceID, notification.Status, orgID, objectType, objectID, destType, destID)
 		}
 
 		err = &ignoredByHandler{"Error in handleGetData: notification.InstanceID != instanceID or notification status is not updated."}
 		SendErrorResponse(writer, err, "", 0)
-	} else if notification.Status == common.Updated || notification.Status == common.Update || notification.Status == common.UpdatePending {
-		//  notification.InstanceID == instanceID
-		updateNotificationRecord = true
-		if trace.IsLogging(logger.DEBUG) {
-			trace.Debug("In handleGetData: notification (status: %s) is updated status,  for %s %s %s %s %s, set updateNotificationRecord to %t \n", notification.Status, orgID, objectType, objectID, destType, destID, updateNotificationRecord)
-		}
+		return
 	} else {
-		// notification status "error" cannot update notification status to "data"
 		if trace.IsLogging(logger.DEBUG) {
-			trace.Debug("In handleGetData: notification (status: %s) is not in updated status,  for %s %s %s %s %s, set updateNotificationRecord to %t \n", notification.Status, orgID, objectType, objectID, destType, destID, updateNotificationRecord)
+			trace.Debug("In handleGetData: notification status is: %s, for %s %s %s %s %s", notification.Status, orgID, objectType, objectID, destType, destID)
 		}
 	}
 
-	if trace.IsLogging(logger.DEBUG) {
-		trace.Trace("Handling object get data, retrieve object data for %s %s\n", objectType, objectID)
+	objectMeta, _, err := Store.RetrieveObjectAndStatus(orgID, objectType, objectID)
+	if err != nil {
+		SendErrorResponse(writer, err, "", 0)
 	}
 
-	if dataReader, err := Store.RetrieveObjectData(orgID, objectType, objectID); err != nil {
+	hasRangeHeader := true
+	startOffset, endOffset, err := common.GetStartAndEndRangeFromRangeHeader(request)
+	if err != nil {
 		SendErrorResponse(writer, err, "", 0)
-	} else {
+	}
+
+	if startOffset == -1 && endOffset == -1 {
+		// Range header not specified, will get all data
+		startOffset = 0
+		endOffset = objectMeta.ObjectSize - 1
+		hasRangeHeader = false
+	}
+
+	dataLength := int(endOffset - startOffset + 1)
+
+	if trace.IsLogging(logger.DEBUG) {
+		trace.Debug("Handling object get data, retrieve object data for %s %s with range %d-%d\n", objectType, objectID, startOffset, endOffset)
+	}
+
+	hasError := false
+	if dataLength == int(objectMeta.ObjectSize) || !hasRangeHeader {
+		dataReader, err := Store.RetrieveObjectData(orgID, objectType, objectID, false)
+		if err != nil {
+			SendErrorResponse(writer, err, "", 0)
+		}
+
 		if dataReader == nil {
 			writer.WriteHeader(http.StatusNotFound)
 		} else {
 			writer.Header().Add("Content-Type", "application/octet-stream")
+			writer.Header().Add("Content-Length", strconv.Itoa(dataLength))
```

Comment:
I think you should delete this line since it will be misleading. The Content-Length header does not get set.. The headers that get set on the responses are:
```
< HTTP/1.1 200 OK
< Date: Thu, 06 Jan 2022 20:12:00 GMT
< Content-Type: application/octet-stream
< Transfer-Encoding: chunked
```

Reply:
@dlarson04 you are right. I will remove this line